functionalDB/chapter.md

```diff
@@ -0,0 +1,1105 @@
+
+(defn ind-at
+   ([db kind] (ind-at db kind (:curr-time db)))
+   ([db kind ts] (kind ((:layers db) ts))))
+````
+## Data behavior and life cycle
+
+So far, our discussion focused on the static aspects of the database - its core constructs and the aggregation of them to a larger structure. It is time now to explore the dynamics of the database - how it is changed throughout time and how the data lifecycle (add => update => remove)  is reflected in it. 
+
+In that regards, and as any archaeologist knows, a piece of data doesn’t actually have a real lifecycle. Once it is created, it exists forever and can only be hidden from the world by a new piece of data in a newer layer. Note for the term ‘hidden’, as it is crucial here. The piece of data has not disappeared and it is not gone, it can be revealed again when the layer it is part of gets exposed. Basically, updating data means adding a layer on top of it with something, and removing data is adding a layer on top of it with nothing. 
+
+This means that when talking about data lifecycle, we really talk about the ongoing changes of the data, as it is exposed in different layers. 
+
+If we go back to our schematic visualization of the database in Figure 1, data lifecycle is simply the way a slice changes by along its layers. 
+
+### The bare necessities
+
+To implement the defined data lifecycle, all we needed to do was to implement three functionalities: 
+
+* adding an entity - an operation that starts with the *add-entity* function
+* removing an entity - an operation that start with the *remove-entity* function
+* updating an entity - an operation that starts with the *update-datom* function. This function is named that way as the update operation is done on a datom (the value of an attribute in an entity) and it is not an entire entity update.
+
+Each of these functionalities, when executed, adds another layer to the database.
+
+#### Adding an entity
+
+The driving force of the entity adding process is the *add-entity* function. It  is responsible to do three things - prepare the entity for addition to the database, add it to the storage and update the indices with the relevant information from the entity.
+
+Preparing an entity means providing it with an id (if it doesn’t have) and setting its timestamp field, these actions take place in the function *fix-new-entity* and it’s auxiliary functions *next-id*, *next-ts* and *update-creation-ts*. These helper functions are responsible for finding the next timestamp of the database and update the creation timestamp of a given entity (which is actually the timestamp field of each of the entity’s attributes), as well as find whether an entity needs to be assigned with an id field, if so set its id and prevent future use of that id.
+
+````clojure
+(defn- next-ts [db] (inc (:curr-time db)))
+
+(defn- update-creation-ts [ent ts-val]
+   (reduce #(assoc-in %1 [:attrs %2 :ts ] ts-val) ent (keys (:attrs ent))))

```

Comment:
Ah, sorry -- I misunderstood the language to mean "I'm setting the _entity's_ ts to the ts of its attributes", not "I'm setting the attribute timestamps to the current".